src/eui/Eui64.cc

```diff
@@ -75,22 +80,99 @@ Eui::Eui64::lookupSlaac(const Ip::Address &c)
 
 // return binary representation of the EUI
 bool
-Eui::Eui64::lookupNdp(const Ip::Address &/*c*/)
+Eui::Eui64::lookupNdp(const Ip::Address &c)
 {
-#if 0 /* no actual lookup coded yet */
-
-    /* no OS yet supported for NDP protocol lookup */
+    bool success = false;
+
+#if _SQUID_LINUX_
+    int rtnetlink_sock;
+    struct {
+        struct nlmsghdr hdr;
+        struct ndmsg ndm;
+    } req;
+    struct sockaddr_nl sa;
+    struct iovec iov;
+    struct msghdr msg;
+    char buf[16384];
+    ssize_t buff_fill;
+    struct nlmsghdr * nl_reply_hdr;
+    struct nlmsghdr * nl_reply_msg;
+    struct rtattr * rt_reply_attr;
+    size_t rt_reply_len;
+    bool found_ip = 0;
+    bool done = 0;
+    struct in6_addr ip;
+    void * mac_ptr;
+
+    c.getInAddr(ip);
+
+    // Prepare rtnetlink request
+    req.hdr.nlmsg_len = NLMSG_LENGTH(sizeof(struct ndmsg));
+    req.hdr.nlmsg_type = RTM_GETNEIGH;
+    req.hdr.nlmsg_flags = NLM_F_REQUEST | NLM_F_DUMP;
+    req.hdr.nlmsg_seq = 0;
+    req.hdr.nlmsg_pid = 0;
+    req.ndm.ndm_family = AF_INET6;
+    req.ndm.ndm_state = 0;
+
+    iov.iov_base = &req;
+    iov.iov_len = req.hdr.nlmsg_len;
+
+    msg.msg_name = &sa;
+    msg.msg_namelen = sizeof(sa);
+    msg.msg_iov = &iov;
+    msg.msg_iovlen = 1;
+    msg.msg_control = NULL;
+    msg.msg_controllen = 0;
+    msg.msg_flags = 0;
+
+    // Send rtnetlink request
+    rtnetlink_sock = socket(AF_NETLINK, SOCK_RAW, NETLINK_ROUTE);
+    if (rtnetlink_sock >= 0) {
```

Comment:
You can reduce scope nesting of all this code by checking for errors instead of success. The if-statement should perform any appropriate debugs() log entry, cleanup and exit.